scripts/dts/edtsdatabase.py

```diff
@@ -0,0 +1,193 @@
+#
+# Copyright (c) 2018 Bobby Noelte
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+from collections.abc import Mapping
+import json
+
+##
+# @brief Extended DTS database
+#
+# Database schema:
+#
+# _edts dict(
+#   'devices':  dict(device-id :  device-struct),
+#   'compatibles':  dict(compatible : sorted list(device-id)),
+#   'device-types': dict(device-type : sorted list(device-id)),
+#   ...
+# )
+#
+# device-struct: dict(
+#   'device-id' : device-id,
+#   'device-type' : list(device-type) or device-type,
+#   'compatible' : list(compatible) or compatible,
+#   'label' : label,
+#   property-name : property-value ...
+# )
+#
+# Database types:
+#
+# device-id: <compatible>_<base address in hex>_<offset in hex>,
+# compatible: any of ['st,stm32-spi-fifo', ...] - 'compatibe' from <binding>.yaml
+# device-type: any of ['GPIO', 'SPI', 'CAN', ...] - 'id' from <binding>.yaml
+# label: any of ['UART_0', 'SPI_11', ...] - label directive from DTS
+#
+class EDTSDatabase(Mapping):
+
+    def __init__(self, *args, **kw):
+        self._edts = dict(*args, **kw)
+        # setup basic database schema
+        for edts_key in ('devices', 'compatibles', 'device-types'):
+            if not edts_key in self._edts:
+                self._edts[edts_key] = dict()
+
+    def __getitem__(self, key):
+        return self._edts[key]
+
+    def __iter__(self):
+        return iter(self._edts)
+
+    def __len__(self):
+        return len(self._edts)
+
+    def convert_string_to_label(self, s):
+        # Transmute ,-@/ to _
+        s = s.replace("-", "_")
+        s = s.replace(",", "_")
+        s = s.replace("@", "_")
+        s = s.replace("/", "_")
+        # Uppercase the string
+        s = s.upper()
+        return s
+
+    def device_id(self, compatible, base_address, offset):
```

Comment:
What is 'offset'?

I think the function should take node_address and compatible into account.
node_address should look like: /soc/i2c@40005800/hts221@5f
From this is should be possible to build device_id: \<compatible>_40005800_5F

And while I don't have any example yet, I think there could be more address level than that, so it should be taken into account upfront: \<compatible>\_<addr_1>_<addr_2>\_<...>\_<addr_n>

Reply:
I took offset only as an assumption from the string you provided as @galak ´s request. You may also call it index, ... Maybe @galak can explain what he would like device id to be build from and look like.

I don´t think it is important how device id will be generated as long as it is unique. My first proposal was to just use the 'label' property as this is unique for all activated devices. 

In my view device_id is just a unique reference to a device struct. It should not be used for other purposes (like generating define labels from it).

Reply:
I'll let @galak provide more information here